### DISCUSSION TITLE
SSL - Microsoft PKI

### DISCUSSION THREAD

**Discussion by christoph-wijns** [ipxe/ipxe#1229] [General]
Hello,

I'm fully aware that my problem is not ipxe related but I need some help.
We are using IPXE for quite a time without no issues.
The one thing that I'm fighting with is SSL. 
Our infrastructure turns on Microsoft IIS & WDS.
Enable Secure Boot encrypt traffic SSL with private certificate.

Sure I'm able to route the traffic https wise with a webserver certificate. But I don't see which certificate I need to bake into snponly.efi.
I can create a Computer Certificate - export pfx and use openssl to extract the necessary parts (TRUST-CERT-KEY) (ex. https://www.ibm.com/docs/en/arl/9.7?topic=certification-extracting-certificate-keys-from-pfx-file). 

After that I'm able to make bin-x86_64-efi/snponly.efi CERT=myca.pem TRUST=myca.pem EMBED=mybootscript.ipxe but obviously that doesn't work ;)

So the main questions are 

- what crosscertificate do I have to create for snponly.efi to negociate with IIS SSL webserver on a Microsoft PKI environment
- which component(s) in pfx (with private key exported) do I need to extract and how that is translated/related to CERT=myca.pem TRUST=myca.pem

Background I'm a MS system engineer set up infrastructure and responsable for OS Deployments at our department but not a GURU @ IPXE & SSL so please be gentle ;)

Thanks

Christoph

![computer_cert](https://github.com/ipxe/ipxe/assets/75322374/760a925b-8111-48a5-812e-e12de1eb83b3)

![computer_cert_2](https://github.com/ipxe/ipxe/assets/75322374/965a807d-ace6-4367-a304-2a5c6fc0d90f)

**Comment by NiKiZe** [comment]
Issues are for bugs, use discussions for questions.

You might benfitt from reading more on how certificate chains usually work.
Do you need a client certificate at all?
If not, make sure you have a certificate for a webbserver on IIS, preferable with a root certificate, take the public part of the root certificate, without the private key, and use that for your trust.

Also make sure this works as expected in a browser.

Even if you are on windows, openssl is a great tool.

**Reply by christoph-wijns** [reply]
Wow first off all many thanks to be ultra responsive !! And my apologies for the opening the issue in issues and not in discussions.

take the public part of the root certificate - can you point me out how to extract this one ?

**You might benefitt from reading more on how certificate chains usually work** - I completely understand your statement but SSL is quite a large knowledge ;)

Many many many thanks

**Reply by NiKiZe** [reply]
This isn't about SSL (which is obsolete and we are now only using TLS) but rather just certificates. The part you need isn't that large, understanding the difference between root, ca, and "issued" and how these are linked together (at the basic level) is quite vital for any tech person today. As well as knowing that there is a public part and a private key.

The certificate on the server needs the private key, the client never sees this and only uses the public key.
When you visit the website in your browser, you will be able to view the certificate used, and export it, this is only the public part, here you will also see the full certificate chain, and can export them, once again of course only the public part.

**Comment by christoph-wijns** [comment]
Hello ;)

I think/thought that I understood your answer so I extracted the root ca & intermediate ca.

![computer_cert_4](https://github.com/ipxe/ipxe/assets/75322374/e5449a6c-6330-47bf-8995-dbc6a5904eed)

![computer_cert_3](https://github.com/ipxe/ipxe/assets/75322374/f9249549-590d-42d0-956b-511632111be4)

Then I baked it with following command

make bin-x86_64-efi/snponly.efi EMBED=ipxescript TRUST=/tmp/rootCA.crt,/tmp/subordinateCA.crt DEBUG=httpcore

Just to be complete - this is an extract of ipxescript

```
#!ipxe
dhcp
chain https://URL/ipxe/Boot/menu.ipxe
exit
```

also tried to add some debugging but SSL is not accepted neither do I have logging.

Many thanks for helping

Christoph

**Reply by christoph-wijns** [reply]
I like to open this thread because I don't arrive.

I have extracted the server certificate from webpage and build the snponly.ipxe.

`make bin-x86_64-efi/snponly.efi EMBED=boot.ipxe CERT=ca.cert.pem TRUST=ca.cert.pem DEBUG=tftp,tls,x509,certstore,open,httpcore`

Secure boot - no answer no logging - secure boot off there is no issue (same snponly.efi)

I know there is tutorial but for me I just don't understand.

Can somebodyplease help.

Christoph

**Reply by NiKiZe** [reply]
Are you using a self signed certificate? If so you will need to trust the root/CA certifcates.

If you want a client certificate you will also need the private key.

**Reply by christoph-wijns** [reply]
@NiKiZe first of all THANKS ;)

That sound obviously and don't get me wrong but previously you said that just the public part is necessary.

Questions

1. Can somebody give me the openssl commands or url to extract the necessary parts from a pfx with private key ?
2. Can somebody  give me thé make command to bake an encrypted snponly.efi for secure boot ?

Like said above I don't want to be lazy but I've tried in the meanwhile so many times and I really don't see it.

Thanks

**Reply by NiKiZe** [reply]
Do you really want to use a https client certificate?

**Reply by christoph-wijns** [reply]
I don't know I want secure boot so I guess that is https ? Or I'm totally wrong - our infrastructure has no access toward internet

**Reply by NiKiZe** [reply]
Secure Boot has nothing to do with https.
But for https, if you don't allow internet access then ipxe must be given trust some other way to accept https connections, you do mention iis and WDS... Do you get any kind of error when using iPXE without the cert stuff at build time?

**Reply by christoph-wijns** [reply]
I've worked with ipxe for years with secure boot off but now with intune I like to enable secure boot and I'm not able establish a connection my first thought was ssl encryption between workstation and IIS but you say that that has nothing to do with secure boot - so can you help me out.
When I enable secure boot and I like to boot generated snponly.efi I get nothing - when I disable secure boot everything works well

Christoph

**Reply by oschroeder8** [reply]
Christoph,

It sounds like there is a misunderstanding how Secure Boot works.

When you add public certificates into the iPXE binary, the way you are doing it, you are essentially telling iPXE which traffic to trust. I.e. your webserver serving your iPXE scripts and images signs it's HTTPS traffic with a private certificate, and you injected the matching public certificate into iPXE, allowing iPXE to ensure that the HTTPS traffic from your webserver is legit.

Secure Boot is the other way round. When you enable secure boot in your firmware, it tells UEFI to check the digitial signature of the iPXE file itself. UEFI firmware contains a set of approved public keys and your iPXE binary must be signed by a matching private key. There are only a few companies that can legitimately sign iPXE binaries, Microsoft being one of them.

There is a way to tell UEFI firmware to trust your own generated public key, but you will have to add that public key to every machine's UEFI - a tedious process. Once your own public key is known to UEFI, you could sign your iPXE binary with the matching private key and secure boot would work.

What you are seeing currently with your unsigned iPXE file is that UEFI checks the file, sees it is not signed by an approved vendor, and then stops booting.

**Reply by NiKiZe** [reply]
So the issue you want to resolve is to have Intune complain less about Secure Boot being disabled?
If so first thing you can do is to enable Secure Boot on your machines, but in factory deployment mode.

If you want to build a signed iPXE, there is a target you can use to sign it do this with test certificate first, and then install that certificate in your firmware.
After this you can get a EV code signing certificate, and then upload that build to MS, this is not something that I would recommend unless you understand code signing and can spend a great deal of time and money on the process.

There is work to have MS sign a shim that would be usable to make signed releases of iPXE without having every release being signed by MS themselves, I would simply wait for that, or perhaps ping MS about them being slow :D

**Comment by christoph-wijns** [comment]
Octavio,

So many thanks for the clear explanation !!!!

In fact you're saying that I'm unable (without to many hassle) let IPXE work smoothly with Secure Boot. 2PintSoftware has a solution but I'd like to do it myself.

I've seen some forums who mentioned the Micrososft signtool.exe - can that be an option ?

Sincerely,

Christoph
________________________________
Van: Octavio Schroeder ***@***.***>
Verzonden: donderdag 7 augustus 2025 15:14
Aan: ipxe/ipxe ***@***.***>
CC: christoph-wijns ***@***.***>; Author ***@***.***>
Onderwerp: Re: [ipxe/ipxe] SSL - Microsoft PKI (Discussion #1229)


Christoph,

It sounds like there is a misunderstanding how Secure Boot works.

When you add public certificates into the iPXE binary, the way you are doing it, you are essentially telling iPXE which traffic to trust. I.e. your webserver serving your iPXE scripts and images signs it's HTTPS traffic with a private certificate, and you injected the matching public certificate into iPXE, allowing iPXE to ensure that the HTTPS traffic from your webserver is legit.

Secure Boot is the other way round. When you enable secure boot in your firmware, it tells UEFI to check the digitial signature of the iPXE file itself. UEFI firmware contains a set of approved public keys and your iPXE binary must be signed by a matching private key. There are only a few companies that can legitimately sign iPXE binaries, Microsoft being one of them.

There is a way to tell UEFI firmware to trust your own generated public key, but you will have to add that public key to every machine's UEFI - a tedious process. Once your own public key is known to UEFI, you could sign your iPXE binary with the matching private key and secure boot would work.

What you are seeing currently with your unsigned iPXE file is that UEFI checks the file, sees it is not signed by an approved vendor, and then stops booting.

**Reply by oschroeder8** [reply]
2PintSoftware went through the lengthy and painful process of submitting their iPXE binary to Microsoft and having them sign it.

Broadcom, through acquisition of VMware and having their Ghost backup software, also has iPXE binaries that are signed:
https://knowledge.broadcom.com/external/article/280113/updated-64bit-ipxeefi-ipxe-v1211+-binari.html

Since I only use the basic snponly.efi and don't need to inject additional certificates, I am using the snponly.efi file from the VMware Autodeploy and that works fine with secure boot.

I don't know the exact process for signing your own compiled iPXE binary. I am sure it is somewhere in the discussions and/or the iPXE documentation pages.
https://ipxe.org/appnote/etoken

**Reply by christoph-wijns** [reply]
@oschroeder8 I can't describe how many thanks you's you get from me with that clear answer you gave me !!!

Vmware AutoDeploy snponly.efi does not work out of the box I suppose. I mean just download the zip packet, extract the snponly64.efi and use it ?

Christoph

**Reply by christoph-wijns** [reply]
Allright i'm getting somewhere.

I've used snponly64.efi.officialkey from Vmware and this works when Secure Boot is active but loops. I don't know/see which boot menu this efi file is calling.

Anyone ;)

Christoph

**Comment by christoph-wijns** [comment]
If I do CRTL+B and 'chain http://...' that works but what do I have to use in boot folder (filename) to automate with snponly64.efi.officialkey ?

**Reply by oschroeder8** [reply]
Oh yeah, sorry. I have been using iPXE for a while now and before they added the option for autoexec.ipxe to get loaded automatically. I guess the VMware version does not support that either.

I am using DHCP to tell iPXE which script file to load. Check out this page for an example:
https://ipxe.org/howto/dhcpd

**Reply by christoph-wijns** [reply]
You're saying DHCP options - Microsoft DHCP ?

**Comment by christoph-wijns** [comment]
Printscreen above writing

<img width="820" height="399" alt="screen-dhcp-01" src="https://github.com/user-attachments/assets/53a32481-ba89-41eb-ab5b-922a66c054c6" />

`#!ipxe
dhcp
chain http://something.be/ipxe/Boot/menu.ipxe
exit`
Where the menu.ipxe is the classic blue screen.

Loop

<img width="1125" height="713" alt="screen-dhcp-02" src="https://github.com/user-attachments/assets/87c84541-2039-440a-ae9a-2e4ee5f1b726" />

My guess is that snponly.efi (vmware) needs other settings in dhcp.

Christoph

**Comment by christoph-wijns** [comment]
Hey,

I've found out that vmware autodeploy uses old gPXE ?

I see when I download the tftp folder from VMWARE that is uses a file called tramp ?

Can somebody please explain the cycle ?

Thanks

Christoph

**Reply by oschroeder8** [reply]
Hi Christoph,

From your screenshot above, you can see that the VMware binary is indeed iPXE and not gPXE.

The way PXE works is that:

- Your system boots up (physical or virtual) and loads the firmware (BIOS or UEFI).
- If you set your firmware to PXE boot, and that is your first boot target, the firmware will send out a DHCP request with some additional information. The additional info is stuff like: I am a PXE client, this is the architecture I identify as, etc.
- The DHCP server receives the information and based on some of that additional information received, it can send back extra configuration settings that the PXE client can use. Such as which executable the PXE client should download to bootstrap the system.
- Many DHCP servers allow you to create "templates" of sort that provide different config settings based on either the architecture of the PXE client, or additional feature flags that the PXE client sent.
- The DHCP link I had previously shows an example of using an older version of ISC's dhcpd to do the following:
  - if a PXE client identifies as a BIOS-based client, direct it to boot the undionly.kpxe iPXE client
  - if a PXE client identifies as a 64-bit UEFI-based client, direct it to boot the snponly.efi iPXE client
  - if the PXE client identifies as iPXE already, then tell it to boot the menu script

This results in the following cycle:
- a 64-bit UEFI instance boots up
- the firmware sends a DHCP request using the built-in firmware PXE client
- The DHCP server gets the request, sees it is a UEFI instance and tells it to boot snponly.efi
- The instance loads and boot snponly.efi (which is iPXE)
- the iPXE client starts up and send a new DHCP request to the DHCP server
- The DHCP server now sees that the DHCP request comes from iPXE and tells it to boot the menu script
- iPXE receives the DHCP ack and loads the menu script

The newer iPXE versions will try to find an autoexec.ipxe menu script from the same location it downloaded the original iPXE binary from. It is my understanding that the VMware iPXE binary does not do that.

I do not know to what extent Microsoft's DHCP server provides logic to do the above.

**Reply by christoph-wijns** [reply]
It fucking works !!!!

Thanks very much for the information.

I'm going to write a blog about this topic

Christoph
